### PR TITLE
Introduce BlockchainDataSource

### PR DESCRIPTION
**Pull request by MrStahlfelge**
This PR attempts to decouple BlockchainContext and its datasource. 
For this, a new BlockchainDataSource class is introduced declaring available data fetch calls.

Decisions that are to be discussed:
* I've made lib-api depend on java-client-generated to use model classes - as an alternative, we would have to make copying interfaces
* ExplorerAndPoolUnspentBoxesLoader still fetches data directly from Retrofit classes. This can be changed by adding more methods to BlockchainDataSource, but before doing the effort I wanted to check if the general approach is accepted
* NodeAndExplorerDataSource has public getters for the API clients. This is made by intention, so that clients needing to call methods from these APIs don't need to build an own, second retrofit service.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -185,7 +185,7 @@ lazy val common = (project in file("common"))
     )
 
 lazy val libApi = (project in file("lib-api"))
-    .dependsOn(common % allConfigDependency)
+    .dependsOn(javaClientGenerated % allConfigDependency, common % allConfigDependency)
```
Comment (aslesarenko) [inline build.sbt:188]: Interfaces should be abstract and can only depend on common module. If the changes require other dependencies, then there is something wrong with the changes.

File: lib-api/src/main/java/org/ergoplatform/appkit/BlockchainDataSource.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.ergoplatform.restapi.client.BlockHeader;
+import org.ergoplatform.restapi.client.ErgoTransaction;
+import org.ergoplatform.restapi.client.ErgoTransactionOutput;
+import org.ergoplatform.restapi.client.NodeInfo;
+import org.ergoplatform.restapi.client.WalletBox;
+
+import java.util.List;
+
+public interface BlockchainDataSource {
```
Comment (aslesarenko) [inline lib-api/src/main/java/org/ergoplatform/appkit/BlockchainDataSource.java:11]: Please add JavaDoc for the interface describing its role and responsibilty.

File: lib-api/src/main/java/org/ergoplatform/appkit/BlockchainDataSource.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
+package org.ergoplatform.appkit;
+
+import org.ergoplatform.restapi.client.BlockHeader;
+import org.ergoplatform.restapi.client.ErgoTransaction;
+import org.ergoplatform.restapi.client.ErgoTransactionOutput;
+import org.ergoplatform.restapi.client.NodeInfo;
+import org.ergoplatform.restapi.client.WalletBox;
```
Comment (aslesarenko) [inline lib-api/src/main/java/org/ergoplatform/appkit/BlockchainDataSource.java:7]: This is not good. 
Those classes and their structure is very specific to node implementation. 
The naming is a bit chaotic and not always consistent because API undergone many iterations over time.
It is often hard to decide what should be used and when.
The model classes are also bound to API methods and often duplicated.

Appkit interfaces on the other hand should be well thought.
It is ok if there is not one-to-one correspondence between Appkit data model and Node API+ExplorerAPI data models.
Moreover, Appkit should have single data model of all blockchain data available through either Node API or Explorer API or whatever API.

If something is missing in existing Appkit interfaces, then can be added either new method/property or new interface representing the corresponding conceptual entity.
So, basically, first conceptual model of Appkit should be extended on the level of interfaces, and only then those interfaces can be implemented in `impl` package in one way or another.

**Review by aslesarenko — APPROVED** [review]

**Review by aslesarenko — CHANGES_REQUESTED** [review]
Please see comments

File: lib-impl/src/main/java/org/ergoplatform/appkit/impl/NodeAndExplorerDataSource.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+import java.util.List;
+
+import jline.internal.Nullable;
+import okhttp3.OkHttpClient;
+import retrofit2.Call;
+import retrofit2.Response;
+import retrofit2.Retrofit;
+import special.sigma.Header;
+
+public class NodeAndExplorerDataSource implements BlockchainDataSource {
```
Comment (aslesarenko) [inline lib-impl/src/main/java/org/ergoplatform/appkit/impl/NodeAndExplorerDataSource.java:42]: Please describe how this implementation implements the interface. Also suggest to rename the class so it ends with Impl

File: common/src/main/java/org/ergoplatform/appkit/JavaHelpers.scala [hunk 5]
```diff
@@ -309,6 +309,11 @@ object JavaHelpers {
     CPreHeader(h.version, h.parentId, h.timestamp, h.nBits, h.height, h.minerPk, h.votes)
   }
 
+  def toHeader(h: Header): special.sigma.Header = {
+    CHeader(h.id, h.version, h.parentId, h.ADProofsRoot, h.stateRoot, h.transactionsRoot, h.timestamp,
+      h.nBits, h.height, h.extensionRoot, h.minerPk, h.powOnetimePk, h.powNonce, h.powDistance, h.votes)
+  }
```
[the remark this answers is not in the document]
Comment (MrStahlfelge) [reply common/src/main/java/org/ergoplatform/appkit/JavaHelpers.scala:315]: Looks like I got lost in Scala!

File: lib-impl/src/main/java/org/ergoplatform/appkit/impl/BlockchainContextImpl.java [hunk 6]
```diff
@@ -1,48 +1,50 @@
...
+    private final List<BlockHeaderImpl> _headers;
     private ErgoWalletImpl _wallet;
 
     public BlockchainContextImpl(
-            ApiClient client, Retrofit retrofit,
-            ExplorerApiClient explorer, Retrofit retrofitExplorer,
+            BlockchainDataSource dataSource,
             NetworkType networkType,
-            NodeInfo nodeInfo, List<BlockHeader> headers) {
+            BlockchainParameters blockchainParameters,
```
[the remark this answers is not in the document]
Comment (MrStahlfelge) [reply lib-impl/src/main/java/org/ergoplatform/appkit/impl/BlockchainContextImpl.java:36]: Former code did it this way, will move.

File: lib-impl/src/main/java/org/ergoplatform/appkit/impl/BlockchainContextImpl.java [hunk 7]
```diff
@@ -52,7 +54,7 @@ public PreHeaderBuilder createPreHeader() {
 
     @Override
     public SignedTransaction signedTxFromJson(String json) {
-        Gson gson = getApiClient().getGson();
+        Gson gson = JSON.createGson().create();
```
[the remark this answers is not in the document]
Comment (MrStahlfelge) [reply lib-impl/src/main/java/org/ergoplatform/appkit/impl/BlockchainContextImpl.java:57]: I thought the old implementation did it this way to save multiple instantiations of Gson. Fetching data and serialization is related, but there is no need to have a data source define Json serialization - it is a standard. Gson is an implementation of this standard and should not leek into the data source interface.

File: lib-impl/src/main/java/org/ergoplatform/appkit/impl/BlockchainContextImpl.java [hunk 8]
```diff
@@ -63,15 +65,16 @@ public UnsignedTransactionBuilder newTxBuilder() {
...
     @Override
     public InputBox[] getBoxesById(String... boxIds) throws ErgoClientException {
         List<InputBox> list = new ArrayList<>();
         for (String id : boxIds) {
-            ErgoTransactionOutput boxData = ErgoNodeFacade.getBoxById(_retrofit, id);
-            if (boxData == null) {
-                throw new ErgoClientException("Cannot load UTXO box " + id, null);
-            }
-            list.add(new InputBoxImpl(this, boxData));
+            list.add(_dataSource.getBoxById(id));
```
[the remark this answers is not in the document]
Comment (MrStahlfelge) [reply lib-impl/src/main/java/org/ergoplatform/appkit/impl/BlockchainContextImpl.java:77]: Former implementation swallowed exceptions on fetching data and returned `null`. Now `ErgoClientException` is generated on `NodeAndExplorerDataSource.executeCall`

File: lib-impl/src/main/java/org/ergoplatform/appkit/impl/BlockchainContextImpl.java [hunk 9]
```diff
@@ -86,72 +89,28 @@ public ErgoProverBuilder newProverBuilder() {
...
+    public BlockchainParameters getParameters() {
+        return _blockchainParameters;
     }
 
     public org.ergoplatform.appkit.PreHeader getPreHeader() {
         return _preHeader;
     }
 
-    public List<BlockHeader> getHeaders() {
+    public List<BlockHeaderImpl> getHeaders() {
```
[the remark this answers is not in the document]
Comment (MrStahlfelge) [reply lib-impl/src/main/java/org/ergoplatform/appkit/impl/BlockchainContextImpl.java:101]: Changing it

File: lib-impl/src/main/java/org/ergoplatform/appkit/impl/ExplorerAndPoolUnspentBoxesLoader.java [hunk 10]
```diff
@@ -39,14 +40,13 @@ public ExplorerAndPoolUnspentBoxesLoader withAllowChainedTx(boolean allowChained
 
     @Override
     public void prepare(@Nonnull BlockchainContext ctx, List<Address> addresses, long grossAmount, @Nonnull List<ErgoToken> tokensToSpend) {
-        if (!(ctx instanceof BlockchainContextImpl)) {
-            throw new IllegalArgumentException("This loader needs to be used with BlockchainContextImpl");
+        if (!(ctx.getDataSource() instanceof NodeAndExplorerDataSource)) {
+            throw new IllegalArgumentException("This loader needs to be used with NodeAndExplorerDataSource");
         }
```
[the remark this answers is not in the document]
Comment (MrStahlfelge) [reply lib-impl/src/main/java/org/ergoplatform/appkit/impl/ExplorerAndPoolUnspentBoxesLoader.java:45]: Plan is to move the needed methods to BlockchainDataSource when everything else is okay with this PR. We can move the class to lib-api then.

File: lib-api/src/main/java/org/ergoplatform/appkit/BlockchainParameters.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+    /**
+     * @return cost of storing 1 byte in UTXO for four years, in nanoErgs
+     */
+    int getStorageFeeFactor();
+
+    /**
+     * @return cost of a transaction output, in computation unit
+     */
+    int getMinValuePerByte();
```
[the remark this answers is not in the document]
Comment (MrStahlfelge) [reply lib-api/src/main/java/org/ergoplatform/appkit/BlockchainParameters.java:22]: Looks like description was wrong

File: lib-impl/src/main/java/org/ergoplatform/appkit/impl/ExplorerAndPoolUnspentBoxesLoader.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply lib-impl/src/main/java/org/ergoplatform/appkit/impl/ExplorerAndPoolUnspentBoxesLoader.java:45]: No, lib-api shouldn't contain classes, only interfaces.

File: lib-impl/src/main/java/org/ergoplatform/appkit/impl/ScalaBridge.scala [hunk 12]
```diff
@@ -221,6 +221,29 @@ object ScalaBridge {
...
+        height = h.getHeight,
+        extensionRoot = h.getExtensionHash.toColl,
+        minerPk = h.getPowSolutionsPk.toGroupElement,
+        powOnetimePk = h.getPowSolutionsW.toGroupElement,
+        powNonce = h.getPowSolutionsN.toColl,
+        powDistance = SigmaDsl.BigInt(h.getPowSolutionsD),
+        votes = h.getVotes.map(Iso.jbyteToByte.to)
+      )
+
+    override def from(a: Header): org.ergoplatform.appkit.BlockHeader = ???
```
Comment (aslesarenko) [inline lib-impl/src/main/java/org/ergoplatform/appkit/impl/ScalaBridge.scala:244]: Hm, this doesn't look like Iso :)

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: lib-api/src/main/java/org/ergoplatform/appkit/BlockchainDataSource.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+    /**
+     * Get a list of unspent boxes for a wallet set up in data source. When data source is a node,
+     * it is the node's users wallet.
+     *
+     * @param minConfirmations   Minimal number of confirmations (optional)
+     * @param minInclusionHeight Minimal box inclusion height (optional)
+     * @return list of InputBoxes to use for a new transaction
+     */
+    List<InputBox> getWalletUnspentBoxes(int minConfirmations, int minInclusionHeight);
```
[the remark this answers is not in the document]
Comment (MrStahlfelge) [reply lib-api/src/main/java/org/ergoplatform/appkit/BlockchainDataSource.java:49]: I agree, I would happily remove or deprecate it together with its BlockchainContext counterpart. Remove or deprecate?

File: lib-api/src/main/java/org/ergoplatform/appkit/BlockHeader.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+    String getExtensionHash();
+
+    String getPowSolutionsPk();
+
+    String getPowSolutionsW();
+
+    BigInteger getPowSolutionsD();
+
+    String getPowSolutionsN();
```
[the remark this answers is not in the document]
Comment (MrStahlfelge) [reply lib-api/src/main/java/org/ergoplatform/appkit/BlockHeader.java:27]: Not using String would mean to leak some Sigma classes to the public interfaces (which is already done in PreHeader interface) - is this okay?

**Review by aslesarenko — APPROVED** [review]
